### PR TITLE
feat: support query offset

### LINKED ISSUE #408: Support for offset in the QueryRecords endpoint
**Is your feature request related to a problem? Please describe.**
The query records endpoint support a limit parameter, but not an offset. An offset would provide a way to skip the keys or records returned in the response without consuming CPU and bandwidth for the unwanted records.

**Describe the solution you'd like**
Support for an Offset in the QueryRecords endpoint.

**Describe alternatives you've considered**
Always downloading all the records and skipping the first "offset" records. The only issue with this approach is that Open Saves times out trying to retrieve the records when the number is large. We have logical stores that have over 20k records, and the request times out before returning to the client.

**Additional context**

### PR DESCRIPTION
**Pull request by loranger2k**
/kind feat

**What this PR does / Why we need it**:
Support query offset to work around timeout issues.

**Which issue(s) this PR fixes**:
Closes #408 

**Special notes for your reviewer**:
The offset support will be leverage when working on Issue #406

### CODE REVIEW

File: api/open_saves.proto [hunk 1]
```diff
@@ -359,16 +359,19 @@ message QueryRecordsRequest {
...
 
   // the limit of the number of records to return.
   int32 limit = 6;
 
-  // If keys_only is set to true, the server will only return records.key 
+  // If keys_only is set to true, the server will only return records.key
   // and store_keys in the QueryRecordsResponse message.
   bool keys_only = 7;
+
+  // Number of records to skip when performing queries - use with keys_only
```
Comment (hongalex) [inline api/open_saves.proto:373]:
```suggestion
  // Number of records to skip when performing queries - use with keys_only.
```

File: internal/pkg/metadb/metadb_test.go [hunk 2]
```diff
@@ -1262,25 +1263,41 @@ func TestMetaDB_QueryRecords(t *testing.T) {
 
 	metaDB := newMetaDB(ctx, t)
 
+	storeKeys := []string{
+		newRecordKey(),
```
Comment (hongalex) [inline internal/pkg/metadb/metadb_test.go:1267]: These should be `newStoreKey()`

File: internal/pkg/metadb/metadb.go [hunk 3]
```diff
@@ -831,6 +830,9 @@ func (m *MetaDB) QueryRecords(ctx context.Context, req *pb.QueryRecordsRequest)
 	if limit := req.GetLimit(); limit > 0 {
 		query = query.Limit(int(limit))
 	}
+	if offset := req.GetOffset(); offset > 0 {
+		query = query.Offset(int(offset))
```
Comment (hongalex) [inline internal/pkg/metadb/metadb.go:834]: Can we enforce this by placing the offset check after `req.GetKeysOnly`?

File: internal/pkg/metadb/metadb_test.go (hunk 2 above)
Comment (loranger2k) [reply internal/pkg/metadb/metadb_test.go:1267]: Good catch.

File: internal/pkg/metadb/metadb.go (hunk 3 above)
Comment (loranger2k) [reply internal/pkg/metadb/metadb.go:834]: I'd like to keep this implementation as is (temporarily) to enable our current client to pull records within a store that has thousands of entries.
Pros: the client is not timing out waiting for the results or downloading them locally (think about pulling 100 records at index 22000 in a store, which is a scenario we see).
Cons: the client ends up generating get requests for all these records.

I am planning to address this after I introduce a GetMulti endpoint in metadb. With it, I'll be able to enforce that the keys oly are pulled with offset>0, and the GetMulti is used to actually get the record content. The reasoning is to keep this entirely in Open Saves without forcing the client to pull the keys and then call again to get the records it wants.

Comment (hongalex) [reply internal/pkg/metadb/metadb.go:834]: The only issue I have is that removing the ability to query by offset _without_ keys only is a breaking change, even if we explicitly disallow this in the comment. At the end of the day, behavior is different. Could you use this change in a branch until the new endpoint is added?

If not, the only other way I could potentially see this is if we explicitly mark the `Offset` field as experimental and subject to change (don't depend on it). The comment could be like 
```// It is EXPERIMENTAL and subject to change or removal without notice.```

Comment (loranger2k) [reply internal/pkg/metadb/metadb.go:834]: Yes, I see your point about breaking change.

I'll see about introducing a GetMulti function to take care of this optimization in this PR, and keep the behavior as is. I.e. when the client asks for keysOnly, they get the key. When they don't, we still use the offset to query keys internally but also calls a new GetMulti function on the results to returns full records. Thus keeping backward compatibility but also enabling querying full records with an offset (even though if we query by keysOnly internally).

**Review by hongalex — APPROVED** [review]

**Pull request merged**